File: app/src/main/java/com/tk4218/grocerylistr/LoginActivity.java
package com.tk4218.grocerylistr;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.TargetApi;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.app.LoaderManager.LoaderCallbacks;

import android.content.CursorLoader;
import android.content.Loader;
import android.database.Cursor;
import android.net.Uri;
import android.os.AsyncTask;

import android.os.Build;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.text.TextUtils;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.inputmethod.EditorInfo;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.GraphRequest;
import com.facebook.GraphResponse;
import com.facebook.login.LoginResult;
import com.facebook.login.widget.LoginButton;
import com.tk4218.grocerylistr.Database.JSONResult;
import com.tk4218.grocerylistr.Database.PasswordStorage;
import com.tk4218.grocerylistr.Database.QueryBuilder;
import com.tk4218.grocerylistr.Model.ApplicationSettings;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.json.JSONException;
import org.json.JSONObject;

import static android.Manifest.permission.READ_CONTACTS;

public class LoginActivity extends AppCompatActivity implements LoaderCallbacks<Cursor> {

    /**
     * Id to identity READ_CONTACTS permission request.
     */
    private static final int REQUEST_READ_CONTACTS = 0;

    //New Account Activity Intent values
    private static final int REQUEST_CREATE_ACCOUNT = 0;
    private static final int ACCOUNT_CREATED = 1;

    //Facebook Login variables
    private static final String EMAIL = "email";
    private CallbackManager mCallbackManager;

    /**
     * Keep track of the login task to ensure we can cancel it if requested.
     */
    private UserLoginTask mAuthTask = null;

    // UI references.
    private AutoCompleteTextView mUsernameView;
    private EditText mPasswordView;
    private View mProgressView;
    private View mLoginFormView;

    private ApplicationSettings mSettings;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        mSettings = new ApplicationSettings(this);

        //See if already logged in and Username is not blank
        if(mSettings.isLoggedIn() && !mSettings.getUser().equals("")){
            Log.d("LOGIN", "Logged in with username: " + mSettings.getUser());
            goToMainActivity();
        }

        /*------------------------------------------------------------
         * Facebook Login
         *------------------------------------------------------------*/
        mCallbackManager = CallbackManager.Factory.create();
        LoginButton loginButton = findViewById(R.id.facebook_login_button);
        loginButton.setReadPermissions(Arrays.asList(EMAIL));

        // Callback registration
        loginButton.registerCallback(mCallbackManager, new FacebookCallback<LoginResult>() {
            @Override
            public void onSuccess(LoginResult loginResult) {
                GraphRequest request = GraphRequest.newMeRequest(
                        loginResult.getAccessToken(),
                        new GraphRequest.GraphJSONObjectCallback() {
                            @Override
                            public void onCompleted(JSONObject object, GraphResponse response) {
                                try {
                                    String userId = object.getString("id");
                                    String email = object.getString("email");
                                    String name = object.getString("name");

                                    showProgress(true);
                                    FacebookLogin facebookLogin = new FacebookLogin(userId, email, name);
                                    facebookLogin.execute();

                                } catch (JSONException e) {e.printStackTrace();}
                            }
                        });

                Bundle parameters = new Bundle();
                parameters.putString("fields", "id,name,email");
                request.setParameters(parameters);
                request.executeAsync();
            }

            @Override
            public void onCancel() {}
            @Override
            public void onError(FacebookException exception) {}
        });

        // Set up the login form.
        mUsernameView = findViewById(R.id.email);
        populateAutoComplete();

        mPasswordView = findViewById(R.id.password);
        mPasswordView.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView textView, int id, KeyEvent keyEvent) {
                if (id == EditorInfo.IME_ACTION_DONE || id == EditorInfo.IME_NULL) {
                    attemptLogin();
                    return true;
                }
                return false;
            }
        });

        Button mEmailSignInButton = findViewById(R.id.email_sign_in_button);
        mEmailSignInButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                attemptLogin();
            }
        });

        TextView newAccount = findViewById(R.id.button_new_account);
        newAccount.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(LoginActivity.this, NewAccountActivity.class);
                startActivityForResult(intent, REQUEST_CREATE_ACCOUNT);
            }
        });

        mLoginFormView = findViewById(R.id.login_form);
        mProgressView = findViewById(R.id.login_progress);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        mCallbackManager.onActivityResult(requestCode, resultCode, data);

        if(requestCode == REQUEST_CREATE_ACCOUNT){
            if(resultCode == ACCOUNT_CREATED){
                String username = data.getStringExtra("Username");
                String firstName = data.getStringExtra("FirstName");
                String lastName = data.getStringExtra("LastName");
                mSettings.login(username, firstName, lastName);
                goToMainActivity();
            }
        }
    }

    private void populateAutoComplete() {
        if (!mayRequestContacts()) {
            return;
        }

        getLoaderManager().initLoader(0, null, this);
    }

    private boolean mayRequestContacts() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        }
        if (checkSelfPermission(READ_CONTACTS) == PackageManager.PERMISSION_GRANTED) {
            return true;
        }
        if (shouldShowRequestPermissionRationale(READ_CONTACTS)) {
            Snackbar.make(mUsernameView, R.string.permission_rationale, Snackbar.LENGTH_INDEFINITE)
                    .setAction(android.R.string.ok, new View.OnClickListener() {
                        @Override
                        @TargetApi(Build.VERSION_CODES.M)
                        public void onClick(View v) {
                            requestPermissions(new String[]{READ_CONTACTS}, REQUEST_READ_CONTACTS);
                        }
                    });
        } else {
            requestPermissions(new String[]{READ_CONTACTS}, REQUEST_READ_CONTACTS);
        }
        return false;
    }

    /**
     * Callback received when a permissions request has been completed.
     */
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        if (requestCode == REQUEST_READ_CONTACTS) {
            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                populateAutoComplete();
            }
        }
    }

    private void goToMainActivity(){
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        finish();
    }

    /**
     * Attempts to sign in or register the account specified by the login form.
     * If there are form errors (invalid email, missing fields, etc.), the
     * errors are presented and no actual login attempt is made.
     */
    private void attemptLogin() {
        if (mAuthTask != null) {
            return;
        }

        // Reset errors.
        mUsernameView.setError(null);
        mPasswordView.setError(null);

        // Store values at the time of the login attempt.
        String username = mUsernameView.getText().toString();
        String password = mPasswordView.getText().toString();

        boolean cancel = false;
        View focusView = null;


        // Check for a valid password, if the user entered one.
        if (!TextUtils.isEmpty(password) && !isPasswordValid(password)) {
            mPasswordView.setError(getString(R.string.error_invalid_password));
            focusView = mPasswordView;
            cancel = true;
        }

        // Check for a valid email address.
        if (TextUtils.isEmpty(username)) {
            mUsernameView.setError(getString(R.string.error_field_required));
            focusView = mUsernameView;
            cancel = true;
        }

        if (cancel) {
            // There was an error; don't attempt login and focus the first
            // form field with an error.
            focusView.requestFocus();
        } else {
            // Show a progress spinner, and kick off a background task to
            // perform the user login attempt.
            showProgress(true);
            mAuthTask = new UserLoginTask(username, password);
            mAuthTask.execute((Void) null);
        }
    }

    private boolean isPasswordValid(String password) {
        return password.length() > 4;
    }

    /**
     * Shows the progress UI and hides the login form.
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB_MR2)
    private void showProgress(final boolean show) {
        // On Honeycomb MR2 we have the ViewPropertyAnimator APIs, which allow
        // for very easy animations. If available, use these APIs to fade-in
        // the progress spinner.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB_MR2) {
            int shortAnimTime = getResources().getInteger(android.R.integer.config_shortAnimTime);

            mLoginFormView.setVisibility(show ? View.GONE : View.VISIBLE);
            mLoginFormView.animate().setDuration(shortAnimTime).alpha(
                    show ? 0 : 1).setListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    mLoginFormView.setVisibility(show ? View.GONE : View.VISIBLE);
                }
            });

            mProgressView.setVisibility(show ? View.VISIBLE : View.GONE);
            mProgressView.animate().setDuration(shortAnimTime).alpha(
                    show ? 1 : 0).setListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    mProgressView.setVisibility(show ? View.VISIBLE : View.GONE);
                }
            });
        } else {
            // The ViewPropertyAnimator APIs are not available, so simply show
            // and hide the relevant UI components.
            mProgressView.setVisibility(show ? View.VISIBLE : View.GONE);
            mLoginFormView.setVisibility(show ? View.GONE : View.VISIBLE);
        }
    }

    @Override
    public Loader<Cursor> onCreateLoader(int i, Bundle bundle) {
        return new CursorLoader(this,
                // Retrieve data rows for the device user's 'profile' contact.
                Uri.withAppendedPath(ContactsContract.Profile.CONTENT_URI,
                        ContactsContract.Contacts.Data.CONTENT_DIRECTORY), ProfileQuery.PROJECTION,

                // Select only email addresses.
                ContactsContract.Contacts.Data.MIMETYPE +
                        " = ?", new String[]{ContactsContract.CommonDataKinds.Email
                .CONTENT_ITEM_TYPE},

                // Show primary email addresses first. Note that there won't be
                // a primary email address if the user hasn't specified one.
                ContactsContract.Contacts.Data.IS_PRIMARY + " DESC");
    }

    @Override
    public void onLoadFinished(Loader<Cursor> cursorLoader, Cursor cursor) {
        List<String> emails = new ArrayList<>();
        cursor.moveToFirst();
        while (!cursor.isAfterLast()) {
            emails.add(cursor.getString(ProfileQuery.ADDRESS));
            cursor.moveToNext();
        }

        addEmailsToAutoComplete(emails);
    }

    @Override
    public void onLoaderReset(Loader<Cursor> cursorLoader) {

    }

    private void addEmailsToAutoComplete(List<String> emailAddressCollection) {
        //Create adapter to tell the AutoCompleteTextView what to show in its dropdown list.
        ArrayAdapter<String> adapter =
                new ArrayAdapter<>(LoginActivity.this,
                        android.R.layout.simple_dropdown_item_1line, emailAddressCollection);

        mUsernameView.setAdapter(adapter);
    }


    private interface ProfileQuery {
        String[] PROJECTION = {
                ContactsContract.CommonDataKinds.Email.ADDRESS,
                ContactsContract.CommonDataKinds.Email.IS_PRIMARY,
        };

        int ADDRESS = 0;
        int IS_PRIMARY = 1;
    }

    /**
     * Represents an asynchronous login/registration task used to authenticate
     * the user.
     */
    public class UserLoginTask extends AsyncTask<Void, Void, Boolean> {
        private QueryBuilder mQb = new QueryBuilder();

        private final String mUsername;
        private final String mPassword;
        private String mFirstName;
        private String mLastName;

        UserLoginTask(String username, String password) {
            mUsername = username;
            mPassword = password;
        }

        @Override
        protected Boolean doInBackground(Void... params) {

            JSONResult existingUser = mQb.getUserPasswordByUsername(mUsername);
            if(existingUser.getCount() == 0){
                existingUser = mQb.getUserPasswordByEmail(mUsername);
            }

            if(existingUser.getCount() > 0){
                try {
                    mFirstName = existingUser.getString("FirstName");
                    mLastName = existingUser.getString("LastName");
                    return PasswordStorage.verifyPassword(mPassword, existingUser.getString("Password"));
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } else{
                return false;
            }


            return false;
        }

        @Override
        protected void onPostExecute(final Boolean success) {
            mAuthTask = null;
            showProgress(false);

            if (success) {
                mSettings.login(mUsername, mFirstName, mLastName);
                goToMainActivity();
            } else {
                mPasswordView.setError("The username or password is incorrect");
                mPasswordView.requestFocus();
            }
        }

        @Override
        protected void onCancelled() {
            mAuthTask = null;
            showProgress(false);
        }
    }

    private class FacebookLogin extends AsyncTask<Void, Void, Void>{
        private QueryBuilder mQb = new QueryBuilder();

        private String mUserId;
        private String mEmail;
        private String mName;
        private String mFirstName;
        private String mLastName;

        FacebookLogin(String userId, String email, String name){
            mUserId = userId;
            mEmail = email;
            mName = name;
        }

        @Override
        protected Void doInBackground(Void... voids) {
            JSONResult existingUser = mQb.getUserByUsername(mUserId);
            if(existingUser.getCount() == 0){
                existingUser = mQb.getUserByEmail(mEmail);
            }
            if(existingUser.getCount() == 0){
                String[] names = mName.split(" ",2);
                mFirstName = names[0];
                mLastName = names.length > 1 ? names[1] : "";
                mQb.insertUser(mUserId, mEmail, "", mFirstName, mLastName);
            } else {
                mUserId = existingUser.getString("Username");
                mFirstName = existingUser.getString("FirstName");
                mLastName = existingUser.getString("LastName");
            }
            return null;
        }
        @Override
        protected void onPostExecute(Void aVoid) {
            showProgress(false);
            mSettings.login(mUserId, mFirstName, mLastName);
            goToMainActivity();
        }
    }
}

File: app/src/main/java/com/tk4218/grocerylistr/LoginActivity.kt
package com.tk4218.grocerylistr

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.TargetApi
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.app.LoaderManager.LoaderCallbacks

import android.content.CursorLoader
import android.content.Loader
import android.database.Cursor
import android.net.Uri
import android.os.AsyncTask

import android.os.Build
import android.os.Bundle
import android.provider.ContactsContract
import android.util.Log
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.ArrayAdapter
import android.widget.TextView

import com.facebook.CallbackManager
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.GraphRequest
import com.facebook.login.LoginResult
import com.tk4218.grocerylistr.Database.QueryBuilder
import com.tk4218.grocerylistr.model.User

import java.util.ArrayList

import org.json.JSONException

import kotlinx.android.synthetic.main.activity_login.*

interface LoginCallback {
    enum class ErrorField {
        USERNAME, PASSWORD
    }
    fun onLoginSuccess(user: User)
    fun onLoginFailure(errorField: ErrorField, errorId: Int)
}

class LoginActivity : AppCompatActivity(), LoaderCallbacks<Cursor>, LoginCallback {
    private var mFacebookCallback: CallbackManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        //See if already logged in and Username is not blank
        if (User.currentUsername(this).isNotEmpty()) {
            Log.d("LoginActivity", "Logged in with username: ${User.currentUsername(this)}")
            goToMainActivity()
        }

        /*------------------------------------------------------------
         * Facebook Login
         *------------------------------------------------------------*/
        mFacebookCallback = CallbackManager.Factory.create()
        facebook_login_button.setReadPermissions(listOf(EMAIL))
        facebook_login_button.registerCallback(mFacebookCallback, object : FacebookCallback<LoginResult> {
            override fun onSuccess(loginResult: LoginResult) {
                val request = GraphRequest.newMeRequest(
                        loginResult.accessToken
                ) { `object`, _ ->
                    try {
                        val userId = `object`.getString("id")
                        val email = `object`.getString("email")
                        val name = `object`.getString("name")

                        showProgress(true)
                        val facebookLogin = FacebookLogin(userId, email, name)
                        facebookLogin.execute()

                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                }

                val parameters = Bundle()
                parameters.putString("fields", "id,name,email")
                request.parameters = parameters
                request.executeAsync()
            }
            override fun onCancel() {}
            override fun onError(exception: FacebookException) {}
        })

        login_password.setOnEditorActionListener(TextView.OnEditorActionListener { _, id, _ ->
            if (id == EditorInfo.IME_ACTION_DONE || id == EditorInfo.IME_NULL) {
                attemptLogin()
                return@OnEditorActionListener true
            }
            false
        })

        email_sign_in_button.setOnClickListener { attemptLogin() }
        button_new_account.setOnClickListener {
            val intent = Intent(this@LoginActivity, NewAccountActivity::class.java)
            startActivityForResult(intent, REQUEST_CREATE_ACCOUNT)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        mFacebookCallback?.onActivityResult(requestCode, resultCode, data)

        if (requestCode == REQUEST_CREATE_ACCOUNT) {
            if (resultCode == ACCOUNT_CREATED) {
                goToMainActivity()
            }
        }
    }

    private fun goToMainActivity() {
        val intent = Intent(this, MainActivity::class.java)
        startActivity(intent)
        finish()
    }

    override fun onLoginSuccess(user: User) {
        showProgress(false)
        User.setCurrentUser(this, user)
        goToMainActivity()
    }
    override fun onLoginFailure(errorField: LoginCallback.ErrorField, errorId: Int) {
        showProgress(false)
        when(errorField) {
            LoginCallback.ErrorField.USERNAME -> {
                login_username.error = getString(errorId)
                login_username.requestFocus()
            }
            LoginCallback.ErrorField.PASSWORD -> {
                login_password.error = getString(errorId)
                login_password.requestFocus()
            }
        }
    }

    /**
     * Attempts to sign in or register the account specified by the login form.
     * If there are form errors (invalid email, missing fields, etc.), the
     * errors are presented and no actual login attempt is made.
     */
    private fun attemptLogin() {
        // Reset errors.
        login_username.error = null
        login_password.error = null

        showProgress(true)
        User.login(login_username.text.toString(), login_password.text.toString(), this)
    }

    /**
     * Shows the progress UI and hides the login form.
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB_MR2)
    private fun showProgress(show: Boolean) {
        // On Honeycomb MR2 we have the ViewPropertyAnimator APIs, which allow
        // for very easy animations. If available, use these APIs to fade-in
        // the progress spinner.
        val shortAnimTime = resources.getInteger(android.R.integer.config_shortAnimTime)

        login_form.visibility = if (show) View.GONE else View.VISIBLE
        login_form.animate().setDuration(shortAnimTime.toLong()).alpha(
                (if (show) 0 else 1).toFloat()).setListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                login_form.visibility = if (show) View.GONE else View.VISIBLE
            }
        })

        login_progress.visibility = if (show) View.VISIBLE else View.GONE
        login_progress.animate().setDuration(shortAnimTime.toLong()).alpha(
                (if (show) 1 else 0).toFloat()).setListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                login_progress.visibility = if (show) View.VISIBLE else View.GONE
            }
        })
    }

    override fun onCreateLoader(i: Int, bundle: Bundle): Loader<Cursor> {
        return CursorLoader(this,
                // Retrieve data rows for the device user's 'profile' contact.
                Uri.withAppendedPath(ContactsContract.Profile.CONTENT_URI,
                        ContactsContract.Contacts.Data.CONTENT_DIRECTORY), ProfileQuery.PROJECTION,

                // Select only email addresses.
                ContactsContract.Contacts.Data.MIMETYPE + " = ?", arrayOf(ContactsContract.CommonDataKinds.Email
                .CONTENT_ITEM_TYPE),

                // Show primary email addresses first. Note that there won't be
                // a primary email address if the user hasn't specified one.
                ContactsContract.Contacts.Data.IS_PRIMARY + " DESC")
    }

    override fun onLoadFinished(cursorLoader: Loader<Cursor>, cursor: Cursor) {
        val emails = ArrayList<String>()
        cursor.moveToFirst()
        while (!cursor.isAfterLast) {
            emails.add(cursor.getString(ProfileQuery.ADDRESS))
            cursor.moveToNext()
        }

        addEmailsToAutoComplete(emails)
    }

    override fun onLoaderReset(cursorLoader: Loader<Cursor>) {
    }

    private fun addEmailsToAutoComplete(emailAddressCollection: List<String>) {
        //Create adapter to tell the AutoCompleteTextView what to show in its dropdown list.
        val adapter = ArrayAdapter(this@LoginActivity,
                android.R.layout.simple_dropdown_item_1line, emailAddressCollection)

        login_username.setAdapter(adapter)
    }


    private interface ProfileQuery {
        companion object {
            val PROJECTION = arrayOf(ContactsContract.CommonDataKinds.Email.ADDRESS, ContactsContract.CommonDataKinds.Email.IS_PRIMARY)
            val ADDRESS = 0
            val IS_PRIMARY = 1
        }
    }

    private inner class FacebookLogin internal constructor(private var mUserId: String?, private val mEmail: String, private val mName: String) : AsyncTask<Void, Void, Void>() {
        private val mQb = QueryBuilder()
        private var mFirstName: String? = null
        private var mLastName: String? = null

        override fun doInBackground(vararg voids: Void): Void? {
            var existingUser = mQb.getUserByUsername(mUserId)
            if (existingUser.count == 0) {
                existingUser = mQb.getUserByEmail(mEmail)
            }
            if (existingUser.count == 0) {
                val names = mName.split(" ".toRegex(), 2).toTypedArray()
                mFirstName = names[0]
                mLastName = if (names.size > 1) names[1] else ""
                mQb.insertUser(mUserId, mEmail, "", mFirstName, mLastName)
            } else {
                mUserId = existingUser.getString("Username")
                mFirstName = existingUser.getString("FirstName")
                mLastName = existingUser.getString("LastName")
            }
            return null
        }

        override fun onPostExecute(aVoid: Void) {
            showProgress(false)
            //mSettings!!.login(mUserId, mFirstName, mLastName)
            goToMainActivity()
        }
    }

    companion object {
        //New Account Activity Intent values
        private val REQUEST_CREATE_ACCOUNT = 0
        private val ACCOUNT_CREATED = 1

        //Facebook Login variables
        private val EMAIL = "email"
    }
}

